typescript/sdk/src/fee/EvmTokenFeeReader.ts

```diff
@@ -0,0 +1,162 @@
+import { constants } from 'ethers';
+
+import {
+  BaseFee__factory,
+  ERC20__factory,
+  LinearFee__factory,
+  RoutingFee__factory,
+} from '@hyperlane-xyz/core';
+import { Address, WithAddress, assert } from '@hyperlane-xyz/utils';
+
+import { MultiProvider } from '../providers/MultiProvider.js';
+import { ChainName, ChainNameOrId } from '../types.js';
+import { HyperlaneReader } from '../utils/HyperlaneReader.js';
+
+import {
+  BaseTokenFeeConfig,
+  OnchainTokenFeeType,
+  TokenFeeConfig,
+  TokenFeeType,
+  onChainTypeToTokenFeeTypeMap,
+} from './types.js';
+
+type DerivedTokenFeeConfig = WithAddress<TokenFeeConfig>;
+export class EvmTokenFeeReader extends HyperlaneReader {
+  constructor(
+    protected readonly multiProvider: MultiProvider,
+    protected readonly chain: ChainNameOrId,
+  ) {
+    super(multiProvider, chain);
+  }
+
+  async deriveTokenFeeConfig(
+    address: Address,
+    destinations?: number[],
+  ): Promise<DerivedTokenFeeConfig> {
+    const tokenFee = BaseFee__factory.connect(address, this.provider);
+
+    let derivedConfig: DerivedTokenFeeConfig;
+    const onchainFeeType = await tokenFee.feeType();
+    switch (onchainFeeType) {
+      case OnchainTokenFeeType.LinearFee:
+        derivedConfig = await this.deriveLinearFeeConfig(address);
+        break;
+      case OnchainTokenFeeType.ProgressiveFee:
+        derivedConfig = await this.deriveProgressiveFeeConfig(address);
+        break;
+      case OnchainTokenFeeType.RegressiveFee:
+        derivedConfig = await this.deriveRegressiveFeeConfig(address);
+        break;
+      case OnchainTokenFeeType.RoutingFee:
+        assert(
+          destinations,
+          `Destinations required for ${onChainTypeToTokenFeeTypeMap[onchainFeeType]}`,
+        );
+        derivedConfig = await this.deriveRoutingFeeConfig(
+          address,
+          destinations,
+        );
+        break;
+      default:
+        throw new Error(
+          `Unsupported token fee type: ${await tokenFee.feeType()}`,
+        );
```

Comment:
Why do we need to re-fetch the `feeType` here? can't we use the `onchainFeeType` variable?